lib/polyfill/pip.js

```diff
@@ -0,0 +1,245 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+goog.provide('shaka.polyfill.PiP');
+
+goog.require('shaka.log');
+goog.require('shaka.polyfill.register');
+
+
+/**
+ * @namespace shaka.polyfill.PiP
+ *
+ * @summary A polyfill to provide PiP support in Safari.
+ */
+
+
+/**
+ * Install the polyfill if needed.
+ *
+ * @this {HTMLMediaElement}
+ */
+shaka.polyfill.PiP.install = function() {
+  if (!window.HTMLVideoElement) {
+    // Avoid errors on very old browsers.
+    return;
+  }
+
+  let proto = HTMLVideoElement.prototype;
+  if (proto.requestPictureInPicture &&
+    document.exitPictureInPicture) {
+    // No polyfill needed.
+    return;
+  }
+
+  if (proto.webkitSupportsPresentationMode &&
+    typeof proto.webkitSetPresentationMode === 'function') {
+    shaka.log.debug('PiP.install');
+
+    /**
+     * polyfill document.pictureInPictureElement
+     */
+    Object.defineProperty(document, 'pictureInPictureElement', {
+      get() {
+        if (!document.polyfillPictureInPictureElement) {
+          const videoElementList = document.querySelectorAll('video');
+
+          for (let i = 0; i < videoElementList.length; i += 1) {
+            const video =
+              /** @type {!HTMLMediaElement} */ (videoElementList[i]);
+
+            if (video.webkitPresentationMode &&
+              video.webkitPresentationMode === 'picture-in-picture') {
+              document.pictureInPictureElement = video;
+              break;
+            }
+          }
+        }
+
+        return document.polyfillPictureInPictureElement || null;
+      },
+
+      set(value) {
+        if (value === document.polyfillPictureInPictureElement) return;
+
+        if (document.polyfillPictureInPictureElement) {
+          document.polyfillPictureInPictureElement.removeEventListener(
+            'webkitpresentationmodechanged',
+            shaka.polyfill.PiP.updatePictureInPictureElementInDocument,
+          );
+        }
+
+        document.polyfillPictureInPictureElement =
+          /** @type {!HTMLMediaElement} */ (value);
+        if (document.polyfillPictureInPictureElement) {
+          document.polyfillPictureInPictureElement.addEventListener(
+            'webkitpresentationmodechanged',
+            shaka.polyfill.PiP.updatePictureInPictureElementInDocument,
+          );
+        }
+      },
+    });
+
+    /**
+     * polyfill document.pictureInPictureEnabled
+     */
+    document.pictureInPictureEnabled = true;
```

Comment:
Is this always the case?  Are there instances where the API is present, but Safari might not support PiP at runtime for some reason?

Reply:
Okay, sounds good.